### PR TITLE
pam-reattach: init at 1.3

### PR DESCRIPTION
**Pull request by lockejan**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by lockejan** [comment]
I still have to find out how to properly link or put the resulting pam_reattach.so in a location where it can be linked from to.
It needs to be referenced to inside /etc/pam.d/sudo

**Comment by jhvst** [comment]
Shouldn't this be done in two phases? Have the package merged and then do a separate PR on nixos-darwin to do [the same as with touch ID](https://github.com/LnL7/nix-darwin/blob/master/modules/security/pam.nix) but instead use `${pkgs.pam_reattach}/lib/pam/pam_reattach.so` as the filename.

**Comment by lockejan** [comment]
> Shouldn't this be done in two phases? Have the package merged and then do a separate PR on nixos-darwin to do [the same as with touch ID](https://github.com/LnL7/nix-darwin/blob/master/modules/security/pam.nix) but instead use `${pkgs.pam_reattach}/lib/pam/pam_reattach.so` as the filename.

Thanks for your comment. You are completely right. Back then I couldn't wrap my head around it and I didn't want to open that PR without further testing. Will try this now locally with nix-darwin and mark that PR as ready once I have a good enough solution.

**Comment by jhvst** [comment]
Gotcha, makes sense. Also, I noticed that [people already have overlays for pam_reattach](https://github.com/LnL7/nix-darwin/pull/228#issuecomment-1212769786). So, this seems good as the latter part is already done!

**Comment by lockejan** [comment]
Well, I just tested it and it works as expected. The nix-darwin implementation still needs some refinement, but nothing that would need anymore changes for this PR.

File: pkgs/os-specific/darwin/pam-reattach/default.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    "-DENABLE_CLI=ON"
+  ];
+  enableParallelBuilding = true;
+
+  nativeBuildInputs = [ cmake clang openpam ];
+
+  meta = with lib; {
+    homepage = "https://github.com/fabianishere/pam_reattach";
+    description =
+      "Reattach to the user's GUI session on macOS during authentication (for Touch ID support in tmux)";
```
Comment (SuperSandro2000) [inline pkgs/os-specific/darwin/pam-reattach/default.nix:31]:
```suggestion
    description = "Reattach to the user's GUI session on macOS during authentication (for Touch ID support in tmux)";
```

File: pkgs/os-specific/darwin/pam-reattach/default.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    license = licenses.mit;
+    maintainers = with maintainers; [ lockejan ];
+    platforms = platforms.unix;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/darwin/pam-reattach/default.nix:34]:
```suggestion
    platforms = platforms.darwin;
```

File: pkgs/os-specific/darwin/pam-reattach/default.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      if stdenv.hostPlatform.system == "x86_64-darwin" then
+        "x86_64"
+      else
+        "arm64"
+    }"
+    "-DENABLE_CLI=ON"
+  ];
+  enableParallelBuilding = true;
+
+  nativeBuildInputs = [ cmake clang openpam ];
```
Comment (SuperSandro2000) [inline pkgs/os-specific/darwin/pam-reattach/default.nix:26]: Isn't openpam not a library and goes to buildInputs?

File: pkgs/os-specific/darwin/pam-reattach/default.nix [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    "-DCMAKE_BUILD_TYPE=Release"
+    "-DCMAKE_OSX_ARCHITECTURES=${
+      if stdenv.hostPlatform.system == "x86_64-darwin" then
+        "x86_64"
+      else
+        "arm64"
+    }"
+    "-DENABLE_CLI=ON"
+  ];
+  enableParallelBuilding = true;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/darwin/pam-reattach/default.nix:24]:
```suggestion
```
enabled by cmake already

File: pkgs/os-specific/darwin/pam-reattach/default.nix [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+  src = fetchFromGitHub {
+    owner = "fabianishere";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "1k77kxqszdwgrb50w7algj22pb4fy5b9649cjb08zq9fqrzxcbz7";
+  };
+
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE=Release"
```
Comment (SuperSandro2000) [inline pkgs/os-specific/darwin/pam-reattach/default.nix:15]:
```suggestion
```
set by cmake hook

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/darwin/pam-reattach/default.nix (hunk 4 above)
Comment (lockejan) [reply pkgs/os-specific/darwin/pam-reattach/default.nix:24]: Didn't know that. Nice 👍

File: pkgs/os-specific/darwin/pam-reattach/default.nix (hunk 5 above)
Comment (lockejan) [reply pkgs/os-specific/darwin/pam-reattach/default.nix:15]: Haha, nice. TIL again.

File: pkgs/os-specific/darwin/pam-reattach/default.nix (hunk 3 above)
Comment (lockejan) [reply pkgs/os-specific/darwin/pam-reattach/default.nix:26]: Not sure about that one. Can you please rephrase this without double negative?😬😃
Are you suggesting it's not necessary during build- but just runtime?🤔

Comment (lockejan) [reply pkgs/os-specific/darwin/pam-reattach/default.nix:26]: Okay, tried it. Works. Good catch.

**Comment by lockejan** [comment]
Ready to merge? @SuperSandro2000

**Comment by SuperSandro2000** [comment]
@ofborg build pam-reattach

**Comment by SuperSandro2000** [comment]
```
-- Build files have been written to: /tmp/nix-build-pam_reattach-1.3.drv-0/source/build
cmake: enabled parallel building
building
build flags: -j1 SHELL=/nix/store/1bsjl5incfnszv7scdh4d02sh45vw2w1-bash-5.1-p16/bin/bash
[ 16%] Building C object CMakeFiles/reattach.dir/src/reattach.c.o
In file included from /tmp/nix-build-pam_reattach-1.3.drv-0/source/src/reattach.c:26:
/tmp/nix-build-pam_reattach-1.3.drv-0/source/src/xpc.h:29:10: fatal error: 'xpc/xpc.h' file not found
#include <xpc/xpc.h>
         ^~~~~~~~~~~
1 error generated.
make[2]: *** [CMakeFiles/reattach.dir/build.make:76: CMakeFiles/reattach.dir/src/reattach.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:87: CMakeFiles/reattach.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

**Comment by lockejan** [comment]
> ```
> -- Build files have been written to: /tmp/nix-build-pam_reattach-1.3.drv-0/source/build
> cmake: enabled parallel building
> building
> build flags: -j1 SHELL=/nix/store/1bsjl5incfnszv7scdh4d02sh45vw2w1-bash-5.1-p16/bin/bash
> [ 16%] Building C object CMakeFiles/reattach.dir/src/reattach.c.o
> In file included from /tmp/nix-build-pam_reattach-1.3.drv-0/source/src/reattach.c:26:
> /tmp/nix-build-pam_reattach-1.3.drv-0/source/src/xpc.h:29:10: fatal error: 'xpc/xpc.h' file not found
> #include <xpc/xpc.h>
>          ^~~~~~~~~~~
> 1 error generated.
> make[2]: *** [CMakeFiles/reattach.dir/build.make:76: CMakeFiles/reattach.dir/src/reattach.c.o] Error 1
> make[1]: *** [CMakeFiles/Makefile2:87: CMakeFiles/reattach.dir/all] Error 2
> make: *** [Makefile:136: all] Error 2
> ```

Should work now. At least I could build it successfully under both platforms.
If there are any improvements regarding the expressions, please let me know.

**Comment by lockejan** [comment]
@SuperSandro2000 next try?😃😉

**Comment by RaitoBezarius** [comment]
@ofborg build pam-reattach

File: pkgs/os-specific/darwin/pam-reattach/default.nix [hunk 6]
```diff
@@ -0,0 +1,45 @@
+{ lib, stdenv, fetchFromGitHub, cmake, openpam, darwin }:
+
+let
+  sdk =
```
Comment (SuperSandro2000) [inline pkgs/os-specific/darwin/pam-reattach/default.nix:4]: cleaned this up in #205509

**Pull request merged**